p2p/transport/webrtc/stream_write.go

```diff
@@ -0,0 +1,204 @@
+	defer w.stream.wg.Done()
+
+	if !w.stream.stateHandler.AllowWrite() {
+		return 0, io.ErrClosedPipe
+	}
+
+	// Check if there is any message on the wire. This is used for control
+	// messages only when the read side of the stream is closed
+	if w.stream.stateHandler.State() == stateReadClosed {
+		w.readLoopOnce.Do(func() {
+			w.stream.wg.Add(1)
+			go func() {
+				defer w.stream.wg.Done()
+				// zero the read deadline, so read call only returns
+				// when the underlying datachannel closes or there is
+				// a message on the channel
+				w.stream.rwc.(*datachannel.DataChannel).SetReadDeadline(time.Time{})
+				var msg pb.Message
+				for {
+					if w.stream.stateHandler.Closed() {
+						return
+					}
+					err := w.stream.reader.state.readMessageFromDataChannel(&msg)
+					if err != nil {
+						if errors.Is(err, io.EOF) {
+							w.stream.close(true, true)
+						}
+						return
+					}
+					if msg.Flag != nil {
+						state, reset := w.stream.stateHandler.HandleInboundFlag(msg.GetFlag())
+						if state == stateClosed {
+							log.Debug("closing: after handle inbound flag")
+							w.stream.close(reset, true)
+						}
+					}
+				}
+			}()
+		})
+	}
+
+	const chunkSize = maxMessageSize - protoOverhead - varintOverhead
+
+	var (
+		err error
+		n   int
+	)
+
+	for len(b) > 0 {
+		end := len(b)
+		if chunkSize < end {
+			end = chunkSize
+		}
+
+		written, err := w.writeMessage(&pb.Message{Message: b[:end]})
+		n += written
+		if err != nil {
+			return n, err
+		}
+		b = b[end:]
+	}
+	return n, err
+}
+
+func (w *webRTCStreamWriter) writeMessage(msg *pb.Message) (int, error) {
+	w.stream.wg.Add(1)
+	defer w.stream.wg.Done()
+
+	// if the next message will add more data than we are willing to buffer,
+	// block until we have sent enough bytes to reduce the amount of data buffered.
+	timeout := make(chan struct{})
+	var deadlineTimer *time.Timer
+
+	for {
+		if !w.stream.stateHandler.AllowWrite() {
+			return 0, io.ErrClosedPipe
+		}
+		// prepare waiting for writeAvailable signal
+		// if write is blocked
+		deadlineUpdated := w.deadlineUpdated.Wait()
+		writeAvailable := w.writeAvailable.Wait()
+
+		writeDeadline, hasWriteDeadline := w.getWriteDeadline()
+		if hasWriteDeadline {
+			// check if deadline exceeded
+			if writeDeadline.Before(time.Now()) {
+				return 0, os.ErrDeadlineExceeded
+			}
+
+			if deadlineTimer == nil {
+				deadlineTimer = time.AfterFunc(time.Until(writeDeadline), func() { close(timeout) })
+				defer deadlineTimer.Stop()
+			}
+			deadlineTimer.Reset(time.Until(writeDeadline))
+		}
+
+		bufferedAmount := int(w.stream.rwc.(*datachannel.DataChannel).BufferedAmount())
+		addedBuffer := bufferedAmount + varintOverhead + proto.Size(msg)
+		if addedBuffer > maxBufferedAmount {
+			select {
+			case <-timeout:
+				return 0, os.ErrDeadlineExceeded
+			case <-writeAvailable:
+				err := w.writeMessageToWriter(msg)
+				if err != nil {
+					return 0, err
+				}
+				return int(len(msg.Message)), nil
```

Comment:
Nit: deduplicate by using a `continue` here and removing the `else` below.

Reply:
No it wouldn't work as we would be waiting for a condition that doesn't exist.